🐛 Bug Report: Missing required config value at 'backend.baseUrl' in 'relative-resolver'

### 📜 Description

I recently upgraded from 1.32.4 to 1.33.0

After a deployment, the root url fails with the following error in the console

```
Error: Missing required config value at 'backend.baseUrl' in 'relative-resolver'
    at $t.getString (reader.esm.js:268:13)
    at Object.factory (apis.esm.js:10:20)
    at ot.load (ApiResolver.esm.js:48:25)
    at ot.loadDeps (ApiResolver.esm.js:57:26)
    at ot.load (ApiResolver.esm.js:47:23)
    at ot.get (ApiResolver.esm.js:33:17)
    at r (useApi.esm.js:18:25)
    at gt (SignInPage.esm.js:63:19)
    at ki (react-dom.production.min.js:160:137)
    at na (react-dom.production.min.js:289:337)
```

If I navigate to the `/catalog` page however, the site functions as normail

### 👍 Expected behavior

No error should be thrown

### 👎 Actual Behavior with Screenshots

I get the following error

```
react-dom.production.min.js:188 Error: Missing required config value at 'backend.baseUrl' in 'relative-resolver'
    at $t.getString (reader.esm.js:268:13)
    at Object.factory (apis.esm.js:10:20)
    at ot.load (ApiResolver.esm.js:48:25)
    at ot.loadDeps (ApiResolver.esm.js:57:26)
    at ot.load (ApiResolver.esm.js:47:23)
    at ot.get (ApiResolver.esm.js:33:17)
    at r (useApi.esm.js:18:25)
    at gt (SignInPage.esm.js:63:19)
    at ki (react-dom.production.min.js:160:137)
    at na (react-dom.production.min.js:289:337)
```

It appears as though the configuration isn't loaded at the time that `SignInPage` renders

    
    

### 👟 Reproduction steps

Locally this works, it's only broken on deployment

### 📃 Provide the context for the Bug.

_No response_

### 🖥️ Your Environment

```
> yarn backstage-cli info
OS:   Windows_NT 10.0.22621 - win32/x64
node: v22.11.0
yarn: 4.4.1
cli:  0.29.1 (installed)
backstage:  1.33.1

Dependencies:
  @backstage/app-defaults                                          1.5.13
  @backstage/backend-app-api                                       1.0.2
  @backstage/backend-common                                        0.25.0
  @backstage/backend-defaults                                      0.5.3
  @backstage/backend-dev-utils                                     0.1.5
  @backstage/backend-openapi-utils                                 0.3.0
  @backstage/backend-plugin-api                                    1.0.1, 1.0.2
  @backstage/catalog-client                                        1.7.1, 1.8.0
  @backstage/catalog-model                                         1.7.0, 1.7.1
  @backstage/cli-common                                            0.1.14, 0.1.15
  @backstage/cli-node                                              0.2.10
  @backstage/cli                                                   0.29.1
  @backstage/config-loader                                         1.9.1, 1.9.2
  @backstage/config                                                1.2.0, 1.3.0
  @backstage/core-app-api                                          1.15.2
  @backstage/core-compat-api                                       0.3.2
  @backstage/core-components                                       0.16.0
  @backstage/core-plugin-api                                       1.10.1
  @backstage/e2e-test-utils                                        0.1.1
  @backstage/errors                                                1.2.4, 1.2.5
  @backstage/eslint-plugin                                         0.1.10
  @backstage/frontend-app-api                                      0.10.1
  @backstage/frontend-defaults                                     0.1.2
  @backstage/frontend-plugin-api                                   0.9.1
  @backstage/frontend-test-utils                                   0.2.2
  @backstage/integration-aws-node                                  0.1.12, 0.1.13
  @backstage/integration-react                                     1.2.1
  @backstage/integration                                           1.15.1, 1.15.2
  @backstage/plugin-api-docs                                       0.12.0
  @backstage/plugin-app-backend                                    0.4.0
  @backstage/plugin-app-node                                       0.1.27
  @backstage/plugin-app                                            0.1.2
  @backstage/plugin-auth-backend-module-atlassian-provider         0.3.2
  @backstage/plugin-auth-backend-module-auth0-provider             0.1.2
  @backstage/plugin-auth-backend-module-aws-alb-provider           0.3.0
  @backstage/plugin-auth-backend-module-azure-easyauth-provider    0.2.2
  @backstage/plugin-auth-backend-module-bitbucket-provider         0.2.2
  @backstage/plugin-auth-backend-module-bitbucket-server-provider  0.1.2
  @backstage/plugin-auth-backend-module-cloudflare-access-provider 0.3.2
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.3.2
  @backstage/plugin-auth-backend-module-github-provider            0.2.2
  @backstage/plugin-auth-backend-module-gitlab-provider            0.2.2
  @backstage/plugin-auth-backend-module-google-provider            0.2.2
  @backstage/plugin-auth-backend-module-guest-provider             0.2.2
  @backstage/plugin-auth-backend-module-microsoft-provider         0.2.2
  @backstage/plugin-auth-backend-module-oauth2-provider            0.3.2
  @backstage/plugin-auth-backend-module-oauth2-proxy-provider      0.2.2
  @backstage/plugin-auth-backend-module-oidc-provider              0.3.2
  @backstage/plugin-auth-backend-module-okta-provider              0.1.2
  @backstage/plugin-auth-backend-module-onelogin-provider          0.2.2
  @backstage/plugin-auth-backend                                   0.24.0
  @backstage/plugin-auth-node                                      0.5.3, 0.5.4
  @backstage/plugin-auth-react                                     0.1.8
  @backstage/plugin-bitbucket-cloud-common                         0.2.25
  @backstage/plugin-catalog-backend-module-azure                   0.2.4
  @backstage/plugin-catalog-backend-module-logs                    0.1.4
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.2.2
  @backstage/plugin-catalog-backend                                1.28.0
  @backstage/plugin-catalog-common                                 1.1.1
  @backstage/plugin-catalog-graph                                  0.4.12
  @backstage/plugin-catalog-import                                 0.12.6
  @backstage/plugin-catalog-node                                   1.14.0
  @backstage/plugin-catalog-react                                  1.14.1
  @backstage/plugin-catalog                                        1.25.0
  @backstage/plugin-events-node                                    0.4.5
  @backstage/plugin-kubernetes-backend                             0.19.0
  @backstage/plugin-kubernetes-common                              0.9.0
  @backstage/plugin-kubernetes-node                                0.2.0
  @backstage/plugin-kubernetes-react                               0.5.0
  @backstage/plugin-kubernetes                                     0.12.0
  @backstage/plugin-org                                            0.6.32
  @backstage/plugin-permission-backend-module-allow-all-policy     0.2.2
  @backstage/plugin-permission-backend                             0.5.51
  @backstage/plugin-permission-common                              0.8.1, 0.8.2
  @backstage/plugin-permission-node                                0.8.5
  @backstage/plugin-permission-react                               0.4.28
  @backstage/plugin-proxy-backend                                  0.5.8
  @backstage/plugin-scaffolder-backend-module-azure                0.2.2
  @backstage/plugin-scaffolder-backend-module-bitbucket-cloud      0.2.2
  @backstage/plugin-scaffolder-backend-module-bitbucket-server     0.2.2
  @backstage/plugin-scaffolder-backend-module-bitbucket            0.3.2
  @backstage/plugin-scaffolder-backend-module-gerrit               0.2.2
  @backstage/plugin-scaffolder-backend-module-gitea                0.2.2
  @backstage/plugin-scaffolder-backend-module-github               0.5.2
  @backstage/plugin-scaffolder-backend-module-gitlab               0.6.1
  @backstage/plugin-scaffolder-backend                             1.27.0
  @backstage/plugin-scaffolder-common                              1.5.7
  @backstage/plugin-scaffolder-node                                0.6.0
  @backstage/plugin-scaffolder-react                               1.14.0
  @backstage/plugin-scaffolder                                     1.27.0
  @backstage/plugin-search-backend-module-catalog                  0.2.5
  @backstage/plugin-search-backend-module-pg                       0.5.38
  @backstage/plugin-search-backend-module-techdocs                 0.3.2
  @backstage/plugin-search-backend-node                            1.3.5
  @backstage/plugin-search-backend                                 1.7.0
  @backstage/plugin-search-common                                  1.2.15
  @backstage/plugin-search-react                                   1.8.2
  @backstage/plugin-search                                         1.4.19
  @backstage/plugin-signals-react                                  0.0.7
  @backstage/plugin-techdocs-backend                               1.11.2
  @backstage/plugin-techdocs-common                                0.1.0
  @backstage/plugin-techdocs-module-addons-contrib                 1.1.17
  @backstage/plugin-techdocs-node                                  1.12.13
  @backstage/plugin-techdocs-react                                 1.2.10
  @backstage/plugin-techdocs                                       1.11.1
  @backstage/plugin-user-settings-common                           0.0.1
  @backstage/plugin-user-settings                                  0.8.15
  @backstage/release-manifests                                     0.0.11
  @backstage/test-utils                                            1.7.1
  @backstage/theme                                                 0.6.1
  @backstage/types                                                 1.1.1, 1.2.0
  @backstage/version-bridge                                        1.0.10
  ```

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, but I'm happy to collaborate on a PR with someone else

--- a/.changeset/great-buttons-cross.md
+++ b/.changeset/great-buttons-cross.md
@@ -1,0 +1,5 @@
+---
+'@backstage/plugin-app-backend': patch
+---
+
+Fixed a bug where config would not be injected on the `/` and `/index.html` paths.

--- a/plugins/app-backend/src/service/appPlugin.test.ts
+++ b/plugins/app-backend/src/service/appPlugin.test.ts
@@ -75,6 +75,7 @@ describe('appPlugin', () => {
       'package.json': '{}',
       dist: {
         static: {},
+        'index.html': '<html><head></head></html>',
         'index.html.tmpl': '<html><head></head></html>',
       },
     });
@@ -91,6 +92,26 @@ describe('appPlugin', () => {
         }),
       ],
     });
+
+    const rootContent = await fetch(`http://localhost:${server.port()}`).then(
+      res => res.text(),
+    );
+
+    expect(rootContent).toBe(`<html><head>
+<script type="backstage.io/config">
+[]
+</script>
+</head></html>`);
+
+    const indexContent = await fetch(
+      `http://localhost:${server.port()}/index.html`,
+    ).then(res => res.text());
+
+    expect(indexContent).toBe(`<html><head>
+<script type="backstage.io/config">
+[]
+</script>
+</head></html>`);
 
     const htmlContent = await fetch(
       `http://localhost:${server.port()}/api/app/some/html5/route`,

--- a/plugins/app-backend/src/service/router.ts
+++ b/plugins/app-backend/src/service/router.ts
@@ -302,7 +302,17 @@ async function createEntryPointRouter({
   staticRouter.use(notFoundHandler());
 
   router.use('/static', staticRouter);
-  router.use(
+
+  const rootRouter = Router();
+  rootRouter.use((req, _res, next) => {
+    // Make sure / and /index.html are handled by the HTML5 route below
+    if (req.url === '/' || req.url === '/index.html') {
+      next('router');
+    } else {
+      next();
+    }
+  });
+  rootRouter.use(
     express.static(rootDir, {
       setHeaders: (res, path) => {
         // The Cache-Control header instructs the browser to not cache html files since it might
@@ -315,7 +325,9 @@ async function createEntryPointRouter({
       },
     }),
   );
+  router.use(rootRouter);
 
+  // HTML5 routing
   router.get('/*', (_req, res) => {
     if (injectResult?.indexHtmlContent) {
       res.setHeader('Content-Type', 'text/html; charset=utf-8');

